### PR TITLE
persistent connection unit test

### PR DESCRIPTION
**Pull request by hs3180**
persistent connection feature(#306) and its integration test.

### CODE REVIEW

**Comment by davvd** [comment]
@hs3180 Thanks for your pull request, let me find someone who can review it

**Comment by davvd** [comment]
@dmzaytsev it's yours, please review

**Comment by dmzaytsev** [comment]
@hs3180 plese fix qulice issues

**Comment by hs3180** [comment]
@dmzaytsev 
I am sorry for that. The qulice issues was fixed.
But there are some strange results from Travis CI.
It failed for the first time https://travis-ci.org/yegor256/takes/jobs/82016143
And got time out for the second time https://travis-ci.org/yegor256/takes/jobs/82032807

**Comment by pecko** [comment]
@hs3180 

> But there are some strange results from Travis CI.
> It failed for the first time https://travis-ci.org/yegor256/takes/jobs/82016143
> And got time out for the second time https://travis-ci.org/yegor256/takes/jobs/82032807

Maybe related to https://github.com/jcabi/jcabi-github/issues/1165

**Comment by hs3180** [comment]
@pecko 
thank you for your help.
I rebuild the project, and get the same result as my first try.
I believe that my patch leading some thing unstable into the project. I will go further to figure out the potential reason.

**Comment by dmzaytsev** [comment]
@hs3180 I would suggest you implement only unit tests here. Full implementation will be too big for one PR.
In my opinion it should be something like this
```
 new FtRemote(new TkText()).exec(
            new FtRemote.Script() {
                @Override
                public void exec(final URI home) throws IOException {
```
you have to parse `home`, extract host and port and do same as in your `KeepAliveITCase`
I think it should be unit tests for `BkBasic`.

Let's implement tests for various scenarios for a persistent connection:
* the server returns or not returns  'Keep-Alive' header depending on the presence  'Keep-Alive' header in the request
* the socket reused if 'Keep-Alive' header exists in the request
* the socket closed if 'Keep-Alive' header not exists in the request
* the socket closed if 'Keep-Alive' header exists in the request on timeout
* another scenario?

@yegor256 are you agree?

**Comment by dmzaytsev** [comment]
> I believe that my patch leading some thing unstable into the project. I will go further to figure out the potential reason.

@hs3180 I think you're right

**Comment by yegor256** [comment]
@dmzaytsev yes, I agree. I'm in general in favor of small pull requests

**Comment by hs3180** [comment]
@dmzaytsev 
great suggestion.
according to http://tools.ietf.org/html/rfc2616#section-14.10, `connection` field is not required in HTTP 1.1(there is no **MUST** statement says that `connection` field is required).
according to http://www.w3.org/Protocols/rfc2616/rfc2616-sec8.html, 
> A significant difference between HTTP/1.1 and earlier versions of HTTP is that persistent connections are the default behavior of any HTTP connection. That is, unless otherwise indicated, the client SHOULD assume that the server will maintain a persistent connection, even after error responses from the server.

so I think there are only two scenario to be tested: 

1. connection is reused by default. establish a connection and handle two set of request-response by the same connection.
1. a client that doesn't support persistent connection. `connection: close` must be given by client and the server should close the connection after response.

**Comment by davvd** [comment]
@ypshenychka review this ticket please, for compliance with our [quality rules](http://www.teamed.io/qa.html)

**Comment by ypshenychka** [comment]
@hs3180 Why is this PR not merged? Have you created new PRs instead of this?

**Comment by hs3180** [comment]
i've rollback my commits and ready to send a smaller pull request. i'm sorry for that github closed my PR after rollback.
            在2015年09月27日 18:08，Yuliya Pshenychka 写道:@hs3180 Why is this PR not merged? Have you created new PRs instead of this?

**Comment by vbodak** [comment]
@hs3180 I see, thanks.
Please correct your last message by indicating an addressee at the beginning.

**Comment by ypshenychka** [comment]
@hs3180 I see, thanks.
Please correct your last message by indicating an addressee at the beginning.

**Comment by ypshenychka** [comment]
@dmzaytsev 
According to our [QA Rules](http://www.teamed.io/qa.html):
>The code reviewer found at least three problems in the code.

Only one issue was found during code review. Try to find at least three problems while future reviews.
Please confirm that everything is clear and you have no questions regarding this approach.

**Comment by dmzaytsev** [comment]
@ypshenychka I confirm

**Comment by ypshenychka** [comment]
@dmzaytsev Thanks,

**Comment by hs3180** [comment]
@dmzaytsev 
I've finished the unit tests described above. But it can't pass the CI obviously since the feature hasn't been implemented.

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 1]
```diff
@@ -60,9 +74,9 @@ public void handlesSocket() throws IOException {
         Mockito.when(socket.getInputStream()).thenReturn(
             new ByteArrayInputStream(
                 Joiner.on("\r\n").join(
-                    "GET / HTTP/1.1",
+                    "POST / HTTP/1.1",
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @hs3180 why did you change this line?

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 2]
```diff
@@ -60,9 +74,9 @@ public void handlesSocket() throws IOException {
...
                     "Host:localhost",
-                    "Content-Length: 2",
+                    "Content-Length: 4",
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @hs3180 and this one?

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 3]
```diff
@@ -108,4 +122,116 @@ public void exec(final URI home) throws IOException {
             }
         );
     }
+    
+    /**
+     * Try to send 2 requests in 1 tcp connection.
+     * @throws IOException
+     */
+    @Test
+    public void persistenceConnection() throws IOException {
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @hs3180 I see only one request in this test. How do you check that's the connection not closed after first request?

**Comment by dmzaytsev** [comment]
@hs3180 3 comments above

File: src/test/java/org/takes/http/BkBasicTest.java (hunk 1 above)
Comment (hs3180) [inline src/test/java/org/takes/http/BkBasicTest.java]: @dmzaytsev 
In most practical applications, GET method shouldn't be attached with a body.
see latest discussion in #306 for more information.

File: src/test/java/org/takes/http/BkBasicTest.java (hunk 2 above)
Comment (hs3180) [inline src/test/java/org/takes/http/BkBasicTest.java]: The request body "hi" is actually "hi\r\n".

File: src/test/java/org/takes/http/BkBasicTest.java (hunk 3 above)
Comment (hs3180) [inline src/test/java/org/takes/http/BkBasicTest.java]: @dmzaytsev 
There are actually two requests.
The last assert statement checks whether two "Hello world" have replied by server.
Maybe it is a bit hard to read, I will refactor it.

**Comment by hs3180** [comment]
@dmzaytsev 
updated for qulice and readability issue.

**Comment by ypshenychka** [comment]
@davvd Quality is acceptable.

**Comment by davvd** [comment]
> @davvd Quality is acceptable.

@ypshenychka thanks for the QA review, we'll work better next time

**Comment by dmzaytsev** [comment]
@davvd @yegor256 this PR has not been reviewed yet

**Comment by hs3180** [comment]
@davvd @ypshenychka 
I'm sorry for closing this PR unexpectly.
this PR is still updating and @dmzaytsev is reviewing now.

**Comment by davvd** [comment]
> @davvd @yegor256 this PR has not been reviewed yet
> 

@dmzaytsev OK, go ahead

**Comment by davvd** [comment]
> @davvd @ypshenychka 
> I'm sorry for closing this PR unexpectedly.
> this PR is still updating and @dmzaytsev is reviewing now.

@hs3180 OK, got it

File: src/test/java/org/takes/http/BkBasicTest.java [hunk 4]
```diff
@@ -108,4 +179,91 @@ public void exec(final URI home) throws IOException {
...
+            InetAddress.getLocalHost()
+        );
+        Mockito.when(socket.getLocalPort()).thenReturn(0);
+        Mockito.when(socket.getInetAddress()).thenReturn(
+            InetAddress.getLocalHost()
+        );
+        Mockito.when(socket.getPort()).thenReturn(0);
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        Mockito.when(socket.getOutputStream()).thenReturn(baos);
+        new BkBasic(new TkGreedy(new TkText(HELLO_WORLD))).accept(socket);
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkBasicTest.java]: @hs3180 I still not understand how you check that the socket still open after first request...

**Comment by dmzaytsev** [comment]
@hs3180 one comment/question above
also please check out this article http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#redundant-constants
> Another typical mistake is to use constants in unit tests to avoid duplicate string/numeric literals in test methods. Don't do this! Every test method should work with its own set of input values.

**Comment by dmzaytsev** [comment]
@hs3180 ping

**Comment by hs3180** [comment]
@dmzaytsev 
thanks for your review. code updated.
I've moved persistent connection feature to a new wrapped Back named `BkReuse`

**Comment by dmzaytsev** [comment]
@hs3180 I don't see any unit test in this PR
I would suggest to start from tests

**Comment by hs3180** [comment]
@dmzaytsev 
so sorry, i've pushed wrong branch.

File: src/test/java/org/takes/http/BkReuseTest.java [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+ * @checkstyle MultipleStringLiteralsCheck (500 lines)
+ */
+@SuppressWarnings("PMD.AvoidDuplicateLiterals")
+public final class BkReuseTest {
+    /**
+     * BkReuse can handle 2 requests in 1 tcp connection.
+     * @throws IOException Socket IOException.
+     */
+    @Test
+    public void handleTwoRequestInOneConnection() throws IOException {
```
Comment (dmzaytsev) [inline src/test/java/org/takes/http/BkReuseTest.java]: @hs3180 `handlesTwoRequestInOneConnection`
http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html?expand=1#test-method-names

**Comment by dmzaytsev** [comment]
@hs3180 let's ignore broken tests and create puzzle to implement missed `BkReuse` 
see http://www.yegor256.com/2009/03/04/pdd.html

**Comment by dmzaytsev** [comment]
@hs3180 ping

**Comment by hs3180** [comment]
@dmzaytsev 
I'm sorry, @davvd ask me to stop working with this issue #306.

**Comment by davvd** [comment]
@dmzaytsev I added 10 mins to @ypshenychka (for QA review) in transaction 66959064

paid, thanks, added **26 mins** to your account, payment `66959076`, 353 hours and 18 mins was spent

review comments (c=11) added as a bonus

+26 added to your rating, current score is: [+1350](http://www.netbout.com/b/35166?open=rating)
